fixes #26548 - graphql: support namespaced modules

timogoebel opened the pull request:
The graphql schema has to contain [three methods to resolve id, types and objects](https://graphql-ruby.org/relay/object_identification.html).

* `id_from_object` converts an object (e.g. a instance of `Host::Managed`) to a global id
*  `object_from_id ` does the opposite: it resolves a global id to an object, basically a `Host::Managed.find(id)`
* `resolve_type` resolves the correcspoding graphql type given an object

These methods did not properly handle namespaced models, e.g. `Host::Managed` or `Subnet::Ipv4` or models from plugins, e.g. `ForemanAnsible::FactName`.

With this commit model classes can now link to the corresponding type by a small DSL:
```ruby
module Host
  class Managed
    graphql_type ::Types::Host
  end
end
```

timogoebel committed:
fixes #26548 - graphql: support namespaced modules

tbrisker commented on app/models/application_record.rb:
```diff
@@ -13,4 +13,12 @@ def self.logger
+    if new_graphql_type
+      @graphql_type = new_graphql_type
+    else
+      @graphql_type || (superclass.respond_to?(:graphql_type) ? superclass.public_send(:graphql_type) : nil)
```
This would recurse up the class tree, either until getting to ApplicationRecord, which will return `nil` because ActiveRecord doesn't respond to `graphql_type`, or return the class instance variable along the way if it was set. Which would happen anyways if the function isn't defined on the child class.
So perhaps we could simplify the implementation greatly by having
```ruby
def self.graphql_type
  nil
end
```
here, and override it only on the classes you want to with something similar to:
```ruby
def self.graphql_type
  ::Types::Host
end
```
and that way we don't need to work with passing lambdas around and calling them in the schema etc

timogoebel replied:
> This would recurse up the class tree, either until getting to ApplicationRecord, which will return `nil` because ActiveRecord doesn't respond to `graphql_type`, or return the class instance variable along the way if it was set.

Yes, that was the aim of this. I don't see a performance problem with this approach to be honest. Isn't that was ruby does unter the hood anyways if you define a method on the superclass?

> override it only on the classes you want to with something similar to:

I really like having a DSL for this. That way it's very active record style. Overriding methods does work, but does not imply `this is a stable API developers can use and rely on`.

> work with passing lambdas around

We just need the lambdas to solve a circular dependency error. The types reference the classes and the classes the types. That did not work in my testing.

tbrisker replied:
Could we just use strings instead of lambdas to prevent the circular dependency? then `safe_constantize` would work for that case and reduce the complexity. I'm trying to avoid lambdas if possible because they make the code more difficult to understand.
i think you can also simplify the retrieval with:
```suggestion
      @graphql_type || superclass.try(:graphql_type)
```

timogoebel replied:
> Could we just use strings instead of lambdas to prevent the circular dependency?

Yes, we definitely can. I actually implemented this in the first iteration but decided to go for lambdas as I think Strings are pretty misleading. I know rails usually supports both strings and lambdas. How about we do the following: I change this to strings and open a follow-up PR that adds supports for lambdas (with proper test coverage etc.). WDYT?

timogoebel replied:
I also want to avoid `try` if possible as this swallows all exceptions which can make debugging harder. I know the code is harder to read but if it helps find a bug where it happens, it's worth it imho.

timogoebel replied:
@tbrisker: What do you think about adding an abstraction layer like this:

```ruby
module Foreman
  module Module
    # This method accepts either a string or a lambda and resolves it to a constant
    def self.resolve(mod)
      mod = mod.call if mod.respond_to?(:call)
      return mod.constantize if mod.instance_of?(String)
      mod
    end
  end
end
```

We can both support strings and lambdas and add proper testing for this.

mmoll commented:
@timogoebel @tbrisker What's the status here? This is marked for 1.22...

timogoebel commented:
@mmoll: I am under the impression that this needs a re-review.

timogoebel commented:
@tbrisker: How can we proceed here? :-)

tbrisker commented on app/graphql/foreman_graphql_schema.rb:
```diff
@@ -20,14 +20,22 @@ def self.object_from_id(id, query_ctx)
-    types[obj.class.name]
+  def self.resolve_type(_, object, _)
+    klass = object.class
+    if klass.respond_to?(:graphql_type) && klass.graphql_type
```
```suggestion
    if klass.try(:graphql_type)
```
Although if you add it to ApplicationRecord, would there be a case where it doesn't respond to `graphql_type`?

timogoebel replied:
Yes, the object could be something different than a descendant of `ApplicationRecord `. E.g. Networks from a Compute Resource or something like this.

tbrisker commented on app/graphql/foreman_graphql_schema.rb:
```diff
@@ -20,14 +20,22 @@ def self.object_from_id(id, query_ctx)
+      return type.call if type.respond_to?(:call)
+      return type.safe_constantize
+    end
+    types[klass.name.demodulize]
```
```
Host::Managed.name.demodulize => "Managed"
```
I'm not sure, is this the expected result in this case?

timogoebel replied:
It also does not work for `Subnet::Ipv4`. However it works for `ForemanPlugin::SomeModel`, that's why I chose `demodulize`. We can simply set `graphql_type` for all namespaced models in core to mitigate this.
But I'm open to suggestions.

timogoebel replied:
I can change this back to `obj.class.name`. All namespaced plugins would need to set `graphql_type` then, which is fine imho. WDYT, @tbrisker?

tbrisker reviewed (changes requested):
See inline, mostly trying to think how we can make this simpler

timogoebel commented:
@tbrisker: Updated the code according to your comments and removed the feature of using lambdas when linking a type to a model. It only accepts a string for now.

tbrisker commented:
[test katello]

tbrisker reviewed (approved):
Sorry for the slow response time, busy days... (not that it's an excuse)

tbrisker commented:
Thanks @timogoebel !
